rust/src/io/external_manifest/commit.rs

```diff
@@ -0,0 +1,401 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Trait for external manifest handler.
+//!
+//! This trait abstracts an external storage with put_if_not_exists semantics.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use object_store::path::Path;
+
+use crate::dataset::{latest_manifest_path, manifest_path};
+use crate::format::{Index, Manifest};
+use crate::io::commit::{CommitError, CommitHandler, ManifestWriter};
+use crate::io::{read_manifest, ObjectStore};
+
+use crate::{Error, Result};
+
+/// External manifest store
+///
+/// This trait abstracts an external storage for source of truth for manifests.
+/// The storge is expected to remember (uri, version) -> manifest_path
+/// and able to run transactions on the manifest_path.
+/// **only supports absolute paths**
+#[async_trait]
+trait ExternalManifestStore: std::fmt::Debug + Send + Sync {
+    /// Get the manifest path for a given uri and version
+    async fn get(&self, uri: &str, version: u64) -> Result<String>;
+
+    /// Get the latest version of a dataset at the path
+    async fn get_latest_version(&self, uri: &str) -> Result<Option<u64>>;
+
+    /// Put the manifest path for a given uri and version, should fail if the version already exists
+    async fn put_if_not_exists(&self, uri: &str, version: u64, path: String) -> Result<()>;
+}
+
+/// External manifest commit handler
+/// This handler is used to commit a manifest to an external store
+/// for detailed design, see https://github.com/lancedb/lance/issues/1183
+#[derive(Debug)]
+struct ExternalManifestCommitHandler {
+    external_manifest_store: Arc<dyn ExternalManifestStore>,
+}
+
+fn make_staging_manifest_path(base: &Path) -> Result<Path> {
+    let id = uuid::Uuid::new_v4().to_string();
+    Path::parse(format!("{base}-{id}")).map_err(|e| Error::IO {
+        message: format!("failed to parse path: {}", e),
+    })
+}
+
+#[async_trait]
+impl CommitHandler for ExternalManifestCommitHandler {
+    /// Get the latest version of a dataset at the path
+    async fn get_latest_version(
+        &self,
+        path: &Path,
+        object_store: &ObjectStore,
+    ) -> std::result::Result<u64, crate::Error> {
+        let version = self
+            .external_manifest_store
+            .get_latest_version(path.as_ref())
+            .await?;
+
+        match version {
+            Some(version) => {
+                let object_store_manifest_path = manifest_path(path, version);
+
+                // external store and object store are out of sync
+                // try to sync them
+                // if sync fails, return error
```

Comment:
Where's the code path for if they are in sync?

Reply:
if we are in sync, just `Ok(version)` this line https://github.com/lancedb/lance/pull/1190/files#diff-d8ab1c2bdfb45c25b964beef780f0f4317a97e5dd891d35777d0509000700863R106